src/czar/CzarRegistry.cc

```diff
@@ -0,0 +1,191 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "czar/CzarRegistry.h"
+
+// System headers
+#include <stdexcept>
+
+// Third party headers
+#include "nlohmann/json.hpp"
+
+// Qserv headers
+#include "cconfig/CzarConfig.h"
+#include "czar/CzarChunkMap.h"
+#include "czar/Czar.h"
+#include "http/Client.h"
+#include "http/Method.h"
+#include "util/common.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+using namespace std;
+using namespace nlohmann;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.czar.CzarRegistry");
+}  // namespace
+
+namespace lsst::qserv::czar {
+
+CzarRegistry::CzarRegistry(std::shared_ptr<cconfig::CzarConfig> const& czarConfig) : _czarConfig(czarConfig) {
+    // Begin periodically updating worker's status in the Replication System's registry
+    // in the detached thread. This will continue before the application gets terminated.
+    thread _registryUpdateThread(&CzarRegistry::_registryUpdateLoop, this);
+    _czarHeartbeatThrd = move(_registryUpdateThread);
+
+    thread _registryWorkerUpdateThread(&CzarRegistry::_registryWorkerInfoLoop, this);
+    _czarWorkerInfoThrd = move(_registryWorkerUpdateThread);
```

Comment:
Same comment for the name of the local variable. You may consider replacing two lines with one:
```
_czarWorkerInfoThrd = move(thread(&CzarRegistry::_registryWorkerInfoLoop, this));
```

Reply:
The second comment regarding these threads is that there is a safer design. Yould could redefine the class as:
```
class CzarRegistry : public std:: enable_shared_from_this <CzarRegistry> {
```
And then create threads in the factory method:
```
    static Ptr create(std::shared_ptr<cconfig::CzarConfig> const& czarConfig) {
        auto ptr = Ptr(new CzarRegistry(czarConfig));
        ptr-> _czarHeartbeatThrd = std::move(std::thread([ptr]() {
            ptr->_registryUpdateLoop();
        });
        ptr->_czarWorkerInfoThrd = std::move(std::thread([ptr]() {
            ptr->_registryWorkerInfoLoop();
        });
        return ptr;
    }
```
The main advantage of this approach is that you no longer need to care about `this` from inside either method `_registryUpdateLoop` or `_registryWorkerInfoLoop` since `ptr` will be captured by threads by copy. The copy of the shared pointer will guarantee the life expectancy of `CzarRegistry` for your threads. This also eliminates any need in `_loop`. See my other comments on this subject downstream.

Reply:
Aside from being able to stop the loop and call join for the threads, I'm also not sure what the exact behavior should be if communications between the czar and workers becomes unstable. The original behavior was to `abort()` which I thought was a bit drastic. However, failed registry contact attempts and failed connections to workers for a few may be the point to terminate the czar. I'm note sure where that line should be drawn.

Reply:
Seeing unstable connections usually is not a good reason for terminating a process. It's not so uncommon to run into those during a startup of a complex distributed application (Qserv, etc.):

1. during the initial start time of the application
2. during rolling upgrades (in Kubernetes)
3. due to intermittent network problems
4. due to crashes and restarts of individual services

Moreover, the main intent here is to make each component as resilient as possible. A decision to abort the application should be left to the human operators.

Good reasons for terminating a process (to count just a few):

1. misconfiguration
2. problems with resources
3. programming bugs triggering assertions, divisions by 0, logic errors, etc.